Eevee.Sleep.Bot/Controllers/Mongo/RoleRecordController.cs

```diff
@@ -0,0 +1,58 @@
+using Eevee.Sleep.Bot.Models;
+using MongoDB.Driver;
+
+namespace Eevee.Sleep.Bot.Controllers.Mongo;
+
+public static class RoleRecordContoller {
+    public static Task SaveRoles(ulong userId, ulong[] roles) {
+        return MongoConst.RoleRecordCollection.UpdateOneAsync(
+            Builders<RoleRecordModel>.Filter.Where(x => x.UserId == userId),
+            Builders<RoleRecordModel>.Update.Set(x => x.Roles, roles), 
+            new UpdateOptions { IsUpsert = true }
+        );
+    }
+
+    public static Task RemoveRoles(ulong userId, ulong[] roles) {
+        return MongoConst.RoleRecordCollection.UpdateOneAsync(
+            Builders<RoleRecordModel>.Filter.Where(x => x.UserId == userId),
+            Builders<RoleRecordModel>.Update.PullAll(x => x.Roles, roles)
+        );
+    }
+
+    public static Task BulkRemoveRoles(ulong[] userIds, ulong[] roles) {
+        return MongoConst.RoleRecordCollection.UpdateManyAsync(
+            Builders<RoleRecordModel>.Filter.In(x => x.UserId, userIds),
+            Builders<RoleRecordModel>.Update.PullAll(x => x.Roles, roles)
+        );
+    }
+
+    public static Task AddRoles(ulong userId, ulong[] roles) {
+        return MongoConst.RoleRecordCollection.UpdateOneAsync(
+            Builders<RoleRecordModel>.Filter.Where(x => x.UserId == userId),
+            Builders<RoleRecordModel>.Update.AddToSetEach(x => x.Roles, roles),
+            new UpdateOptions { IsUpsert = true }
+        );
+    }
+
+    public static Task BulkAddRoles(ulong[] userIds, ulong[] roles) {
+        var bulkOps = new List<WriteModel<RoleRecordModel>>();
+        foreach (var userId in userIds) {
+            bulkOps.Add(
+                new UpdateOneModel<RoleRecordModel>(
+                    Builders<RoleRecordModel>.Filter.Where(r => r.UserId == userId),
+                    Builders<RoleRecordModel>.Update.AddToSetEach(r => r.Roles, roles)
+                ) { IsUpsert = true }
+            );
+        }
```

Comment:
If not bulkOperation, UserId was not correctly numbered and only one record was created.

Reply:
Could you elaborate, maybe give some examples? I would expect `updateMany()` to work because the MongoDB equivalent query of `updateMany({userId: {$in: userIds}}, {$addToSet: roles})` should work, if the intended behavior is to add all `roles` to `userIds`.

Reply:
## When bulk operation
userId is stored
```
        var bulkOps = new List<WriteModel<RoleRecordModel>>();
        foreach (var userId in userIds) {
            bulkOps.Add(
                new UpdateOneModel<RoleRecordModel>(
                    Builders<RoleRecordModel>.Filter.Where(r => r.UserId == userId),
                    Builders<RoleRecordModel>.Update.AddToSetEach(r => r.Roles, roles)
                ) { IsUpsert = true }
            );
        }

        return MongoConst.RoleRecordCollection.BulkWriteAsync(bulkOps);
```
![image](https://github.com/RaenonX-PokemonSleep/eevee-sleep-discord-bot/assets/6148777/ded05a2f-d2b0-4231-8990-e5ef656d1210)

## When updateMany
userId is gone
```
        return MongoConst.RoleRecordCollection.UpdateManyAsync(
            Builders<RoleRecordModel>.Filter.In(x => x.UserId, userIds),
            Builders<RoleRecordModel>.Update.AddToSetEach(x => x.Roles, roles),
            new UpdateOptions { IsUpsert = true }
        );
```
![image](https://github.com/RaenonX-PokemonSleep/eevee-sleep-discord-bot/assets/6148777/f78d9058-7e6b-4c33-8cc0-8398648f98a8)

Reply:
Ah yeah, this is an upsert so it needs to be handled individually, makes sense